src/csharp/Grpc.Core/Internal/AsyncCallBase.cs

```diff
@@ -0,0 +1,407 @@
+        /// </summary>
+        public void CancelWithStatus(Status status)
+        {
+            lock (myLock)
+            {
+                Preconditions.CheckState(started);
+                cancelRequested = true;
+
+                if (!disposed)
+                {
+                    call.CancelWithStatus(status);
+                }
+            }
+        }
+
+        protected void InitializeInternal(CallSafeHandle call)
+        {
+            lock (myLock)
+            {
+                // Make sure this object and the delegated held by it will not be garbage collected
+                // before we release this handle.
+                gchandle = GCHandle.Alloc(this);
+                this.call = call;
+            }
+        }
+
+        /// <summary>
+        /// Initiates sending a message. Only once send operation can be active at a time.
+        /// completionDelegate is invoked upon completion.
+        /// </summary>
+        protected void StartSendMessageInternal(TWrite msg, AsyncCompletionDelegate completionDelegate)
+        {
+            byte[] payload = UnsafeSerialize(msg);
+
+            lock (myLock)
+            {
+                Preconditions.CheckNotNull(completionDelegate, "Completion delegate cannot be null");
+                CheckSendingAllowed();
+
+                call.StartSendMessage(payload, sendFinishedHandler);
+                sendCompletionDelegate = completionDelegate;
+            }
+        }
+
+        /// <summary>
+        /// Requests receiving a next message.
+        /// </summary>
+        protected void StartReceiveMessage()
+        {
+            lock (myLock)
+            {
+                Preconditions.CheckState(started);
+                Preconditions.CheckState(!disposed);
+                Preconditions.CheckState(!errorOccured);
+
+                Preconditions.CheckState(!readingDone);
+                Preconditions.CheckState(!readPending);
+
+                call.StartReceiveMessage(readFinishedHandler);
+                readPending = true;
+            }
+        }
+
+        /// <summary>
+        /// Default behavior just completes the read observer, but more sofisticated behavior might be required
+        /// by subclasses.
+        /// </summary>
+        protected virtual void CompleteReadObserver()
+        {
+            FireReadObserverOnCompleted();
+        }
+
+        /// <summary>
+        /// If there are no more pending actions and no new actions can be started, releases
+        /// the underlying native resources.
+        /// </summary>
+        protected bool ReleaseResourcesIfPossible()
+        {
+            if (!disposed && call != null)
+            {
+                if (halfclosed && readingDone && finished)
+                {
+                    ReleaseResources();
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void ReleaseResources()
```

Comment:
Because cannot release native resources in a destructor.
What you normally do for objects that own unmanaged resources is you make them implement IDisposable https://msdn.microsoft.com/en-us/library/system.idisposable%28v=vs.110%29.aspx

But AsyncCall has to manage its own lifetime anyway (basically waiting for completion callbacks to be delivered), so implementing IDisposable doesn't really help me much here (I had that before and realized it's not helpful).